### DISCUSSION TITLE
[next 14.0.4] Getting Server Actions error from non server actions request/response (Working on Chromium, not for Safari or FireFox)

### DISCUSSION THREAD

**Discussion by prunusnira** [vercel/next.js#59105] [App Router]
### What I want to implement

1. personal authorization via third-party library
2. how this library works?

(1) service: request with authorization request (on same tab or popup) with return URL parameters passed
(2) third party: do authorization process -> open page of return URL with result
(3) service: check result and doing rest job

> problem occurs at (2), where return URL opens

> Please note that chromium browsers (including Chrome, Edge, etc...) works but firefox or safari does not
also all the browsers on iOS does not work (as u might know, every browser on iOS uses Safari engine, not chromium)

### Return component (at current status)

```typescript
'use server';

import ComponentCertResult from '@component/user/cert/componentCertResult';
import { Suspense } from 'react';
import { headers } from 'next/headers';

async function PageCertResult() {
    const header = headers();
    console.log('x-forwarded-host', header.get('x-forwarded-host'));
    console.log('host', header.get('host'));

    return (
        <Suspense fallback={<>Loading Cert Result</>}>
            <ComponentCertResult />
        </Suspense>
    );
}

export default PageCertResult;
```

> What ComponentCertResult do? -> clean up process and close popup or move to result page

### Client Component

```typescript
    const openNiceCert = () => {
        setCertData({
            uuid,
            requestData,
            redirectUrl,
            checkValidationData,
            type: isMobile ? 'mobile' : 'pc',
        });
        if (isMobile) {
            router.push(LINK.AUTH.nicecert);
        } else {
            const popupWidth = 420;
            const popupHeight = 720;
            const popupX = window.screen.width / 2 - popupWidth / 2;
            const popupY = window.screen.height / 3 - popupHeight / 3;
            window.open(
                LINK.AUTH.nicecert,
                '_blank',
                `--- popup properties here ---`,
            );
        }
    };
```

> only executes third party work

### Errors i got and what i did

This is first error message i got from return url

> `x-forwarded-host` header with value `our web page domain` does not match `origin` header with value `domain from third party` from a forwarded Server Actions request. Aborting the action.

After search, i found #58295 #58019 
I found that

- actually, except chromium browsers, x-forwarded-host and origin is different.

So, I
- add experimental.serverActions.allowedOrigins / allowedForwardedHosts in next.config.js

```
    experimental: {
        serverActions: {
            allowedOrigins: [
                'localhost:3000',
                'our-own-url.something.com',
                'nice.checkplus.co.kr', // url from third party
            ],
            allowedForwardedHosts: [
                'localhost:3000',
                'our-own-url.something.com',
                'nice.checkplus.co.kr', // url from third party
            ],
        },
    },
```

* note that including protocol 'http://' or 'https://' in allowedOrigin caused Error: Connection Closed

which made error changed into

> Error: Failed to find Server Action "null". This request might be from an older or newer deployment. Original error: Invariant: Missing 'next-action' header.

So, I added

```typescript
'use server';
```

to server component but it shows same error.

As this is not a server-actions component, what should I do to resolve this?

### Additional information

* I'm currently on Next 14.0.4, but also tried 13.5.6~14.0.3
* nodejs 18.18.2

**Comment by justoverclockl** [comment]
i have the same error, but for now all my component are client

```
Error: Failed to find Server Action "null". This request might be from an older or newer deployment. Original error: Invariant: Missing 'next-action' header.
2|vfgestFE  |     at ry (/var/www/node_modules/next/dist/compiled/next-server/app-page.runtime.prod.js:16:1644)
2|vfgestFE  |     at /var/www/node_modules/next/dist/compiled/next-server/app-page.runtime.prod.js:15:6358
2|vfgestFE  |     at AsyncLocalStorage.run (node:async_hooks:338:14)
2|vfgestFE  |     at rm (/var/www/node_modules/next/dist/compiled/next-server/app-page.runtime.prod.js:15:5711)
2|vfgestFE  |     at rD (/var/www/node_modules/next/dist/compiled/next-server/app-page.runtime.prod.js:16:22552)
2|vfgestFE  |     at /var/www/node_modules/next/dist/compiled/next-server/app-page.runtime.prod.js:16:25097
2|vfgestFE  |     at AsyncLocalStorage.run (node:async_hooks:338:14)
2|vfgestFE  |     at Object.wrap (/var/www/node_modules/next/dist/compiled/next-server/app-page.runtime.prod.js:13:16241)
2|vfgestFE  |     at /var/www/node_modules/next/dist/compiled/next-server/app-page.runtime.prod.js:16:24987
2|vfgestFE  |     at AsyncLocalStorage.run (node:async_hooks:338:14)
```

**Comment by mgarf** [comment]
Same

**Comment by orenaksakal** [comment]
same here
```
Error: Failed to find Server Action "null". This request might be from an older or newer deployment. Original error: Invariant: Missing 'next-action' header.
    at ry (/var/task/node_modules/.pnpm/next@14.0.4_@babel+core@7.23.6_react-dom@18.2.0_react@18.2.0/node_modules/next/dist/compiled/next-server/app-page.runtime.prod.js:16:1644)
    at /var/task/node_modules/.pnpm/next@14.0.4_@babel+core@7.23.6_react-dom@18.2.0_react@18.2.0/node_modules/next/dist/compiled/next-server/app-page.runtime.prod.js:15:6358
    at AsyncLocalStorage.run (node:async_hooks:346:14)
    at rm (/var/task/node_modules/.pnpm/next@14.0.4_@babel+core@7.23.6_react-dom@18.2.0_react@18.2.0/node_modules/next/dist/compiled/next-server/app-page.runtime.prod.js:15:5711)
    at rD (/var/task/node_modules/.pnpm/next@14.0.4_@babel+core@7.23.6_react-dom@18.2.0_react@18.2.0/node_modules/next/dist/compiled/next-server/app-page.runtime.prod.js:16:22552)
    at /var/task/node_modules/.pnpm/next@14.0.4_@babel+core@7.23.6_react-dom@18.2.0_react@18.2.0/node_modules/next/dist/compiled/next-server/app-page.runtime.prod.js:16:25097
    at AsyncLocalStorage.run (node:async_hooks:346:14)
    at Object.wrap (/var/task/node_modules/.pnpm/next@14.0.4_@babel+core@7.23.6_react-dom@18.2.0_react@18.2.0/node_modules/next/dist/compiled/next-server/app-page.runtime.prod.js:13:16241)
    at /var/task/node_modules/.pnpm/next@14.0.4_@babel+core@7.23.6_react-dom@18.2.0_react@18.2.0/node_modules/next/dist/compiled/next-server/app-page.runtime.prod.js:16:24987
    at AsyncLocalStorage.run (node:async_hooks:346:14)
```

**Comment by heapyxyz** [comment]
I have the same issue while using `await fetch()`. Fortunately everything works as intended in my web app. Why is this happening tho?
```
Missing `origin` header from a forwarded Server Actions request.
Error: Failed to find Server Action "null". This request might be from an older or newer deployment. Original error: Invariant: Missing 'next-action' header.
    at ry (/home/heapy/steamlookup/node_modules/next/dist/compiled/next-server/app-page.runtime.prod.js:16:1644)
    at /home/heapy/steamlookup/node_modules/next/dist/compiled/next-server/app-page.runtime.prod.js:15:6358
    at AsyncLocalStorage.run (node:async_hooks:346:14)
    at rm (/home/heapy/steamlookup/node_modules/next/dist/compiled/next-server/app-page.runtime.prod.js:15:5711)
    at rD (/home/heapy/steamlookup/node_modules/next/dist/compiled/next-server/app-page.runtime.prod.js:16:22552)
    at /home/heapy/steamlookup/node_modules/next/dist/compiled/next-server/app-page.runtime.prod.js:16:25097
    at AsyncLocalStorage.run (node:async_hooks:346:14)
    at Object.wrap (/home/heapy/steamlookup/node_modules/next/dist/compiled/next-server/app-page.runtime.prod.js:13:16241)
    at /home/heapy/steamlookup/node_modules/next/dist/compiled/next-server/app-page.runtime.prod.js:16:24987
    at AsyncLocalStorage.run (node:async_hooks:346:14)
```

**Comment by NilsWestgardh** [comment]
Also getting this error:
Error: Failed to find Server Action "null". This request might be from an older or newer deployment. Original error: Invariant: Missing 'next-action' header.

**Comment by lukebelbina** [comment]
Same issue for me

**Comment by LoremakerEgg** [comment]
Also having the same issue.

"Error: Failed to find Server Action "null". This request might be from an older or newer deployment. Original error: Invariant: Missing 'next-action' header."

Unsure how to resolve it.

**Comment by serkanbektas** [comment]
Add this to header and try again. It worked for me:

```
  {
    key: 'Access-Control-Allow-Origin',
    value: process.env.NEXT_PUBLIC_APP_URL,
  },
```

**Reply by Teqwon-Norman** [reply]
Yes, this also worked for me.

I only used the key: 'Access-Control-Allow-Origin'

**Reply by osammy** [reply]
What Header ?

**Reply by j3rryl** [reply]
I'm a newbie, please elaborate which header, or provide code with the fix. Thank you.

**Reply by serkanbektas** [reply]
Sorry for delay.

You also need to define **NEXT_PUBLIC_APP_URL** in your env file.

in your next.config.js file:
```
module.exports = {
    headers: [
        ...
        {
          key: 'Access-Control-Allow-Origin',
          value: process.env.NEXT_PUBLIC_APP_URL,
        },
    ],
}
```

https://nextjs.org/docs/pages/api-reference/next-config-js/headers

**Reply by andy-t-wang** [reply]
I tried this and am still getting the error. Any way to reproduce this issue to debug on local dev? I'm only seeing this after deployment

**Comment by SarfrazAlame** [comment]
same error i got


Error: Failed to find Server Action "c44e2f10789ec4c5fbc4dad37671bce191042615". This request might be from an older or newer deployment. Original error: Invariant: Couldn't find action module ID from module map.
    at t3 (C:\Users\Administrator\OneDrive\Desktop\intagram\twitter\node_modules\next\dist\compiled\next-server\app-page.runtime.dev.js:39:1693)
    at C:\Users\Administrator\OneDrive\Desktop\intagram\twitter\node_modules\next\dist\compiled\next-server\app-page.runtime.dev.js:38:7079
    at AsyncLocalStorage.run (node:async_hooks:346:14)
    at t4 (C:\Users\Administrator\OneDrive\Desktop\intagram\twitter\node_modules\next\dist\compiled\next-server\app-page.runtime.dev.js:38:6387)
    at rk (C:\Users\Administrator\OneDrive\Desktop\intagram\twitter\node_modules\next\dist\compiled\next-server\app-page.runtime.dev.js:39:25940)
    at C:\Users\Administrator\OneDrive\Desktop\intagram\twitter\node_modules\next\dist\compiled\next-server\app-page.runtime.dev.js:39:28648
    at AsyncLocalStorage.run (node:async_hooks:346:14)
    at Object.wrap (C:\Users\Administrator\OneDrive\Desktop\intagram\twitter\node_modules\next\dist\compiled\next-server\app-page.runtime.dev.js:36:16172)
    at C:\Users\Administrator\OneDrive\Desktop\intagram\twitter\node_modules\next\dist\compiled\next-server\app-page.runtime.dev.js:39:28536
    at AsyncLocalStorage.run (node:async_hooks:346:14)
    at Object.wrap (C:\Users\Administrator\OneDrive\Desktop\intagram\twitter\node_modules\next\dist\compiled\next-server\app-page.runtime.dev.js:36:15420)
    at r_ (C:\Users\Administrator\OneDrive\Desktop\intagram\twitter\node_modules\next\dist\compiled\next-server\app-page.runtime.dev.js:39:28463)
    at rW.render (C:\Users\Administrator\OneDrive\Desktop\intagram\twitter\node_modules\next\dist\compiled\next-server\app-page.runtime.dev.js:39:32559)
    at doRender (C:\Users\Administrator\OneDrive\Desktop\intagram\twitter\node_modules\next\dist\server\base-server.js:1394:44)
    at cacheEntry.responseCache.get.routeKind (C:\Users\Administrator\OneDrive\Desktop\intagram\twitter\node_modules\next\dist\server\base-server.js:1543:46)
    at ResponseCache.get (C:\Users\Administrator\OneDrive\Desktop\intagram\twitter\node_modules\next\dist\server\response-cache\index.js:49:26)
    at DevServer.renderToResponseWithComponentsImpl (C:\Users\Administrator\OneDrive\Desktop\intagram\twitter\node_modules\next\dist\server\base-server.js:1463:53)

**Reply by anisharaz** [reply]
i had the same issues while i deploy my nextjs app in docker, i fixed using a session duration in next auth config 
```
session: {
    maxAge: 7 * 24 * 60 * 60,
  },
```

**Comment by teddybee** [comment]
I have the same error in a nextjs 14.1.0 project.

**Comment by marutsuki** [comment]
Yep I'm getting the same problem:
```
Error: Failed to find Server Action "null". This request might be from an older or newer deployment. Original error: Invariant: Missing 'next-action' header.
    at t3 (C:\Users\marut\Documents\Projects\chirper\node_modules\next\dist\compiled\next-server\app-page.runtime.dev.js:39:1693)
    at C:\Users\marut\Documents\Projects\chirper\node_modules\next\dist\compiled\next-server\app-page.runtime.dev.js:38:7079
    at AsyncLocalStorage.run (node:async_hooks:346:14)
    at t4 (C:\Users\marut\Documents\Projects\chirper\node_modules\next\dist\compiled\next-server\app-page.runtime.dev.js:38:6387)
    at rk (C:\Users\marut\Documents\Projects\chirper\node_modules\next\dist\compiled\next-server\app-page.runtime.dev.js:39:25940)
    at C:\Users\marut\Documents\Projects\chirper\node_modules\next\dist\compiled\next-server\app-page.runtime.dev.js:39:28648
    at AsyncLocalStorage.run (node:async_hooks:346:14)
    at Object.wrap (C:\Users\marut\Documents\Projects\chirper\node_modules\next\dist\compiled\next-server\app-page.runtime.dev.js:36:16172)
    at C:\Users\marut\Documents\Projects\chirper\node_modules\next\dist\compiled\next-server\app-page.runtime.dev.js:39:28536
    at AsyncLocalStorage.run (node:async_hooks:346:14)
    at Object.wrap (C:\Users\marut\Documents\Projects\chirper\node_modules\next\dist\compiled\next-server\app-page.runtime.dev.js:36:15420)
    at r_ (C:\Users\marut\Documents\Projects\chirper\node_modules\next\dist\compiled\next-server\app-page.runtime.dev.js:39:28463)
    at rW.render (C:\Users\marut\Documents\Projects\chirper\node_modules\next\dist\compiled\next-server\app-page.runtime.dev.js:39:32559)
    at doRender (C:\Users\marut\Documents\Projects\chirper\node_modules\next\dist\server\base-server.js:1394:44)
    at cacheEntry.responseCache.get.routeKind (C:\Users\marut\Documents\Projects\chirper\node_modules\next\dist\server\base-server.js:1555:34)
    at ResponseCache.get (C:\Users\marut\Documents\Projects\chirper\node_modules\next\dist\server\response-cache\index.js:49:26)
    at DevServer.renderToResponseWithComponentsImpl (C:\Users\marut\Documents\Projects\chirper\node_modules\next\dist\server\base-server.js:1463:53)
    at C:\Users\marut\Documents\Projects\chirper\node_modules\next\dist\server\base-server.js:992:121
    at NextTracerImpl.trace (C:\Users\marut\Documents\Projects\chirper\node_modules\next\dist\server\lib\trace\tracer.js:104:20)
    at DevServer.renderToResponseWithComponents (C:\Users\marut\Documents\Projects\chirper\node_modules\next\dist\server\base-server.js:992:41)
    at DevServer.renderPageComponent (C:\Users\marut\Documents\Projects\chirper\node_modules\next\dist\server\base-server.js:1856:35)
    at async DevServer.renderToResponseImpl (C:\Users\marut\Documents\Projects\chirper\node_modules\next\dist\server\base-server.js:1894:32)
    at async DevServer.pipeImpl (C:\Users\marut\Documents\Projects\chirper\node_modules\next\dist\server\base-server.js:911:25)
    at async NextNodeServer.handleCatchallRenderRequest (C:\Users\marut\Documents\Projects\chirper\node_modules\next\dist\server\next-server.js:271:17)
    at async DevServer.handleRequestImpl (C:\Users\marut\Documents\Projects\chirper\node_modules\next\dist\server\base-server.js:807:17)
    at async C:\Users\marut\Documents\Projects\chirper\node_modules\next\dist\server\dev\next-dev-server.js:331:20
    at async Span.traceAsyncFn (C:\Users\marut\Documents\Projects\chirper\node_modules\next\dist\trace\trace.js:151:20)
    at async DevServer.handleRequest (C:\Users\marut\Documents\Projects\chirper\node_modules\next\dist\server\dev\next-dev-server.js:328:24)
    at async invokeRender (C:\Users\marut\Documents\Projects\chirper\node_modules\next\dist\server\lib\router-server.js:163:21)
    at async handleRequest (C:\Users\marut\Documents\Projects\chirper\node_modules\next\dist\server\lib\router-server.js:342:24)
    at async requestHandlerImpl (C:\Users\marut\Documents\Projects\chirper\node_modules\next\dist\server\lib\router-server.js:366:13)
    at async Server.requestListener (C:\Users\marut\Documents\Projects\chirper\node_modules\next\dist\server\lib\start-server.js:140:13)
```

**Comment by jjang16** [comment]
Getting the same error after deploying to next 14.1.0
We're using customServer.

```
 Error: Failed to find Server Action "null". This request might be from an older or newer deployment. Original error: Invariant: Missing 'next-action' header.
    at rv (/user/handsomejang/app/node_modules/next/dist/compiled/next-server/app-page.runtime.prod.js:16:1666)
    at /user/handsomejang/app/node_modules/next/dist/compiled/next-server/app-page.runtime.prod.js:15:7002
    at AsyncLocalStorage.run (node:async_hooks:346:14)
    at rg (/user/handsomejang/app/node_modules/next/dist/compiled/next-server/app-page.runtime.prod.js:15:6317)
    at rz (/user/handsomejang/app/node_modules/next/dist/compiled/next-server/app-page.runtime.prod.js:16:24715)
    at /user/handsomejang/app/node_modules/next/dist/compiled/next-server/app-page.runtime.prod.js:16:27397
    at AsyncLocalStorage.run (node:async_hooks:346:14)
    at Object.wrap (/user/handsomejang/app/node_modules/next/dist/compiled/next-server/app-page.runtime.prod.js:13:16207)
    at /user/handsomejang/app/node_modules/next/dist/compiled/next-server/app-page.runtime.prod.js:16:27287
    at AsyncLocalStorage.run (node:async_hooks:346:14) 
```

**Reply by alvaromyconnect** [reply]
@jjang16, I have precisely the same issue. Could you let me know if you found a solution?
I tried the Cors headers, but it didn't work.

**Comment by Rafael-Nicolau** [comment]
Same thing here, next 14.1.0, using a component with "use client" and <Script/>:

```
Error: Failed to find Server Action "null". This request might be from an older or newer deployment. Original error: Invariant: Missing 'next-action' header.
```
 
Maybe this is an “internal bug” ?

**Reply by Rafael-Nicolau** [reply]
OK, so I got what happened to me:

Where I work, we need to use some “external” scripts, one of them after being called runs a request: `this.xmlhttp.open('POST', '/service/isLoggedIn.php', true);`

Since next is “handling” routes after it receives this post request, it returns a default 404 error page. 
This somehow is “polluting” something, that then gives the message in our terminal:
`Error: Failed to find Server Action "null". This request might be from an older or newer deployment. Original error: Invariant: Missing 'next-action' header.`

So, the only solution I found, for me, since I can't change this external script, was to actually create a middleware and do a proper re-routing of the request.

So the question I make is, this message was generated in my terminal because of the request was “wrong” (not intended to next to be honest), OR, because when generating the error message next also throws an error?

**Comment by sgeunhi** [comment]
Same Error
next 14.0.4

```Error: Failed to find Server Action "null". This request might be from an older or newer deployment. Original error: Invariant: Missing 'next-action' header.```

**Comment by marutsuki** [comment]
Okay, this was the issue for me personally. I was writing a server actions file. Which then I converted into a standard REST API route. I forgot to remove the "use server" tag.

**Comment by steventkrawczyk** [comment]
Experiencing the same issue

**Comment by itorz7** [comment]
same too

**Comment by M4h45amu7x** [comment]
same too

**Comment by Rieranthony** [comment]
Same issue for me, happens at random in my logs

**Comment by zchwyng** [comment]
Same here...

**Comment by console-logs** [comment]
## Unorthodox Solution
I faced the same issue and I've done the following to fix it in my codebase. 

### How the problem begin
I have multiple server actions defined in my project and each action is defined in its own file. 
For example `userSignupAction` is defined in `userActions.ts` file and `addToCartAction` defined in `cartActions.ts` file.

Everything was working well until one day I decided to move the `userSignupAction` to a different file i.e from `userActions.ts` file to `authActions.ts` file and changed all the references where it was being called. All of a sudden my `userSignupAction` was not working.

### Methods Tried but failed
1. `rm -r .next && npm run dev`
2. `rm -r node_modules && npm i && npm run dev`
3. Made sure next.js is upto date, `14.1.0`
4. Renamed the file and updated import references.
5. Created a new file with a different name.
6. Undo all the changes i.e. back to the original state when things were working.

None of the above worked. Was getting the same error:
`Error: Failed to find Server Action "e5578114f36066c2761f8417a72dca7fc4da09ac". This request might be from an older or newer deployment. Original error: Invariant: Couldn't find action module ID from module map.`

### Solution that worked for me
I just copied `userSignupAction` from `userActions.ts` file and pasted in `cartActions.ts` file and changed all the references where it was being called. It worked! To double check, I undid all the changes and was able to recreate the same error again.

Can some one give this method a try and let me know if that worked for you?

**Comment by ivan-kuzma-scx** [comment]
Same for me, happens randomly in logs.

Error: Failed to find Server Action "null". This request might be from an older or newer deployment. Original error: Invariant: Missing 'next-action' header.

**Reply by batmar125** [reply]
Exactly, same happening here, already in production. Have no idea what is causing those logs. The Pages Router solutions were tried.

**Reply by serkanbektas** [reply]
Did you try https://github.com/vercel/next.js/discussions/59105#discussioncomment-8223381?

**Comment by azizoid** [comment]
same (and only) error causes 16k logs every week :/

**Reply by serkanbektas** [reply]
Did you try https://github.com/vercel/next.js/discussions/59105#discussioncomment-8223381?

**Comment by c4miloarriagada** [comment]
Same error here Next   v.14.0.4

**Reply by serkanbektas** [reply]
Did you try https://github.com/vercel/next.js/discussions/59105#discussioncomment-8223381?

**Reply by c4miloarriagada** [reply]
Thanks man, this works for me (for the moment) . In my next.config.js i added this code

```
 async headers() {
    return [
      {
        source: '/(.*)',
        headers: [
          {
            key: 'Access-Control-Allow-Origin',
            value: process.env.YOUR_PUBLIC_NEXT_URL,
          },
        ],
      },
    ];
  },
 ```

**Comment by KajSzy** [comment]
I've got this error when trying to use nextjs rewrites for Mixpanel analytics
```json
{
  "destination": "https://api-eu.mixpanel.com/:slug",
  "source": "/mp/:slug"
}
```

and in Mixpanel init
```js
api_host: '/mp',
```

**Comment by tcouper** [comment]
I had the same error when doing a redirect from within a NextJS API route on v 14.1.0

The fix for me was to switch `NextResponse.redirect()` to `Response.redirect()`

I hope this helps

**Reply by Emre86** [reply]
Using a redirect with NextResponse in a POST method generates a POST request.
With the Response redirect, it generates a GET request and there are no errors.
It seems that NextResponse uses the original "context" of the request whereas Response creates a new "context".

**Comment by OsamaQureshi147** [comment]
Has anyone got a workaround for this? I have tried adding `headers` and `source` as recommended but still getting the same issue.

**Comment by daithi-coombes** [comment]
I had the same issue, none of the above solved it either.

By monkey patching console.logs around my project I narrowed it down to the `fs` module in `models/index.ts`, once commented out then issue went away. Offending code:

`./models/index.ts`
```typescript
console.log('Loading models...')
// let foo = []
// fs
//   .readdirSync(__dirname + '/models')
//   .filter(file => {
//     console.log('file: ', file)
//     foo.push(file)
//     return (
//       file.indexOf('.') !== 0 &&
//       file !== basename &&
//       file.slice(-3) === '.ts' &&
//       file.indexOf('.test.ts') === -1
//     )
//   })
//   .forEach(file => {
//     const model = require(path.join(__dirname, file))// (sequelize, DataType)
//     db[model.name] = model
//   })
// console.log('foo: ', foo)
```

**Comment by uroskutlaca** [comment]
Hi there. Pretty new to web dev so if you think of a better solution please share. What solved the problem for me was adding "preventDefault()" and "location.reload()".  My next.config.js file is using default settings, I didn't change anything there ( I tried, but it didn't work for me).


```
export default async function AddPayment() {

  const handleSubmit = async (event:any) => {

      event.preventDefault()    --> this one
      location.reload();     --> this one

  try{
    const response = await fetch('http://localhost:3000/api/endpoint', {
      method: 'POST',
      headers: {
    'Content-Type': 'application/json',
  },cache: "no-store",
      body: JSON.stringify({
   ....
    }),
    });
    if (response.ok) {
      console.log('Data successfully sent');
    } else {
      console.error('Error in POST Payment:', response.status, response.statusText);
    }
}catch(error){
      console.log('Error in POST Payment:',error);
    }
};
  return (
....
    )
}
```

**Comment by madfcat** [comment]
Same. 14.0.4
I am getting it in the production.
I do not even use server actions in the whole project. I will try headers solution though.

**Comment by vitordvr** [comment]
Many forget to switch from onSubmit to "action" when using a form; this might help.

**Reply by cderszteler** [reply]
Could you please explain that in more detail?

**Reply by yonattan95** [reply]
Or possibly many changed the action prop to onclic prop of form element. Then you will change type prop of button to "button". This resolve the error for me.

**Comment by Ozzy1997** [comment]
I hade the same exact problem,
moving my (directory/route.ts) to an api file and updating the form action="/api/directory/" solved the problem for me.
i'm doing more test now and trying to understand why this solved it for me
i already empty the cash and rebuild the app multiple times 
adding the api directory was the only thing that worked so far

**Comment by viraniaman94** [comment]
This is happening for me for a completely empty nextjs project with only the default `/` path with the nextjs default page, deployed on ECS fargate. 

Tried the following things to get rid of it but nothing worked:
1. Update nextjs to latest stable version as of today 14.1.3
2. Added `Access-Control-Allow-Origin` header in next.config.mjs

I don't even have any server actions defined yet. It also happens in an infinite loop and creates thousands of log statements saying `Error: Failed to find Server Action "null". This request might be from an older or newer deployment. Original error: Invariant: Missing 'next-action' header`. It is also only happening in production and not happening locally.

Does anyone else have the same issue as me? Has anyone been able to figure this out? This seems like an extremely critical bug to me, why has the NextJS team been ignoring it for 3+ months since it was first reported?

**Reply by jessethomson** [reply]
I haven't tried a completely empty NextJS project, but I am also running on AWS ECS fargate and seeing this same issue.

**Reply by viraniaman94** [reply]
FWIW since I was just starting off a new project and also fairly new to frontend dev in general (with several years of backend exp), I just switched to Remix. Couldn't be happier. Feels like an order of magnitude better framework than NextJS. There are no weird gotchas (that I've come across yet), very well designed and stable API with non overlapping surfaces, and can achieve everything I wanted with nextjs without all the complicated concepts.

**Comment by jasongalvin** [comment]
I was getting that same "Missing 'next-action' header" error with Next 14.1.3.

Then I discovered that I was passing a bad URL to axios (due to a missing env variable) in one of my server actions.  Once I fixed the axios request, the error went away and it works fine.

**Comment by imbhargav5** [comment]
This is probably not the answer for everyone. 
I was using yarn. 
I moved to pnpm. All errors are now gone.

**Reply by ninthsun91** [reply]
I use pnpm, but have this issue. So, this won't apply to everyone.

**Comment by bestcomy-duplicate** [comment]
I use next 14.1.2 App Router. when I was using next-auth v5 to implement user login in my app, I got the error message you guys mentioned here.

> Error: Failed to find Server Action "null". This request might be from an older or newer deployment. Original error: Invariant: Missing 'next-action' header.

I call signIn to do login in my customized form, but I get above error message. It shows "HTTP/1.1 404 Not Found" when I check the http response using Fiddle. I am sure to implement route.ts under /api/auth/[...nextauth] correctly.
I have tried all the workround you guys provided, but no one works for me?
`async function onSubmit(values: z.infer<typeof formSchema>) {
    signIn("credentials", { username: values.username, password: values.password,redirect: false, redirectTo: callbackUrl })
  }`

> HTTP/1.1 404 Not Found
Access-Control-Allow-Origin: http://localhost:3000
Access-Control-Allow-Methods: GET, POST, PUT, DELETE, OPTIONS
Access-Control-Allow-Headers: Content-Type, Authorization
Cache-Control: no-store, must-revalidate
Vary: RSC, Next-Router-State-Tree, Next-Router-Prefetch, Next-Url, Accept-Encoding
X-Powered-By: Next.js
Content-Type: text/html; charset=utf-8
Content-Encoding: gzip
Date: Sun, 31 Mar 2024 10:35:28 GMT
Connection: keep-alive
Keep-Alive: timeout=5
Transfer-Encoding: chunked

**Reply by Teqwon-Norman** [reply]
Not sure if you tried this solution, but I created my authentication options in a separate module and imported it into the route.ts in the /api/auth/[...nextauth] path to handle the get and post requests.

Hope this helps any

**Reply by bestcomy-duplicate** [reply]
It's imported handler to deal with the post request cause  I am using 5.0.0-beta.16
I am going to revert to old version to try.
thanks a lot.

**Comment by AndonMitev** [comment]
Same issue, is not reproducible every time but happens...

```
Error: Failed to find Server Action "null". This request might be from an older or newer deployment. Original error: Invariant: Missing 'next-action' header.
    at ry (/var/task/node_modules/next/dist/compiled/next-server/app-page.runtime.prod.js:16:1666)
    at /var/task/node_modules/next/dist/compiled/next-server/app-page.runtime.prod.js:15:7035
    at AsyncLocalStorage.run (node:async_hooks:338:14)
    at rm (/var/task/node_modules/next/dist/compiled/next-server/app-page.runtime.prod.js:15:6350)
    at rq (/var/task/node_modules/next/dist/compiled/next-server/app-page.runtime.prod.js:18:1255)
    at /var/task/node_modules/next/dist/compiled/next-server/app-page.runtime.prod.js:18:3935
    at AsyncLocalStorage.run (node:async_hooks:338:14)
    at Object.wrap (/var/task/node_modules/next/dist/compiled/next-server/app-page.runtime.prod.js:13:16239)
    at /var/task/node_modules/next/dist/compiled/next-server/app-page.runtime.prod.js:18:3825
    at AsyncLocalStorage.run (node:async_hooks:338:14)
```

And happens only on prod

**Reply by AndonMitev** [reply]
Do you think if we revert back to 14.0.3 or even 14.0.0 for example might solve the issue?

@leerob

**Comment by davidlyashuk** [comment]
Same issue. Next version 14.1.4
`Error: Failed to find Server Action "null". This request might be from an older or newer deployment. Original error: Invariant: Missing 'next-action' header.`

**Reply by Qodestackr** [reply]
> Same issue. Next version 14.1.4 `Error: Failed to find Server Action "null". This request might be from an older or newer deployment. Original error: Invariant: Missing 'next-action' header.`

How did you fix this? We're getting the same issue

**Comment by prunusnira** [comment]
my case is resolved by fixing api usage, but it looks like there's several cases which occurs `server action null` problem.
so, please open thread on your problem. i think there will be different approach for each situation

and it turned out that 'working on chrome/chromium' looks like bug, or chromium handles/modifies request.

**Comment by Yhozen** [comment]
I believe this is related to undici

**Comment by erwanriou** [comment]
In my case i think it's related to internationalization. I highly suspect it doesn't like the `const host = req.headers.get("host")` and maybe i should setup a fallback?

The solution of adding a `Access-Control-Allow-Origin` header would probably not fit for me as it's exposed to multiple domains and putting a wildcard would be the same as before... (you cannot add multiple domains into this header)

```js
import { Domains } from "..."
import createMiddleware from "next-intl/middleware"

const intlMiddleware = createMiddleware({
  locales: Domains?.map(({ defaultLocale }) => defaultLocale),
  defaultLocale: "us",
  domains: Domains,
  localePrefix: "never",
  timeZone: "Europe/Madrid"
})

const middleware = req => {
  // GET DEFAULT LOCALE
  const host = req.headers.get("host")
  const { defaultLocale = "us" } = Domains.find(({ domain }) => domain === host)
  req.cookies.set("NEXT_LOCALE", defaultLocale)

  // APPLY TRANSLATION MIDDLEWARE
  return intlMiddleware(req)
}

export default middleware

export const config = {
  matcher: ["/", "/((?!api|_next|_vercel|.*\\..*).*)"]
}
```

This started appearing after switching to the `app router` first with the library `next-translate` then again with the library `next-intl`

**Comment by anisharaz** [comment]
```
Error: Failed to find Server Action "null". This request might be from an older or newer deployment. Original error: Invariant: Missing 'next-action' header.
    at ry (/app/node_modules/next/dist/compiled/next-server/app-page.runtime.prod.js:16:1666)
    at /app/node_modules/next/dist/compiled/next-server/app-page.runtime.prod.js:15:7035
    at AsyncLocalStorage.run (node:async_hooks:346:14)
    at rm (/app/node_modules/next/dist/compiled/next-server/app-page.runtime.prod.js:15:6350)
    at rq (/app/node_modules/next/dist/compiled/next-server/app-page.runtime.prod.js:18:1255)
    at /app/node_modules/next/dist/compiled/next-server/app-page.runtime.prod.js:18:3935
    at AsyncLocalStorage.run (node:async_hooks:346:14)
    at Object.wrap (/app/node_modules/next/dist/compiled/next-server/app-page.runtime.prod.js:13:16239)
    at /app/node_modules/next/dist/compiled/next-server/app-page.runtime.prod.js:18:3825
    at AsyncLocalStorage.run (node:async_hooks:346:14)
```
### am still getting the error in my log , the discussion is closed but still cannot find a solution, should i consider it as a bug or what am i supposed to do ?

**Comment by erwanriou** [comment]
well i created a new one here since this one is closed --> https://github.com/vercel/next.js/discussions/64214
